src/Infrastructure/stubs/mpiuni/mpi.c

```diff
@@ -179,6 +199,84 @@ int Petsc_MPI_Finalize(void)
   return 0;
 }
 
+int ESMC_MPI_Alltoallw(void *sendbuf, int *sendcounts, int *sdispls,
+                       MPI_Datatype *sendtypes, void *recvbuf, int *recvcounts,
+                       int *rdispls, MPI_Datatype *recvtypes, MPI_Comm comm)
+{
+  // Since we are only implementing this for the single-processor case, the counts, displs
+  // and types arguments should all have length 1. We assume that's the case in this
+  // implementation.
+
+  // Displacements are not implemented so return an error code if they are non-zero
+  if (sdispls[0] != 0 || rdispls[0] != 0) {
+    return MPI_ERR_INTERN;
+  }
+
+  MPIUNI_Memcpy(recvbuf, sendbuf, sendcounts[0]*sendtypes[0], CHECK_FOR_MPI_IN_PLACE_SOURCE);
+  return MPI_SUCCESS;
+}
+
+int ESMC_MPI_Scatterv(void *sendbuf, int *sendcounts, int *displs,
+                      MPI_Datatype sendtype, void *recvbuf, int recvcount,
+                      MPI_Datatype recvtype, int root, MPI_Comm comm)
+{
+  // Since we are only implementing this for the single-processor case, the sendcounts and
+  // displs arguments should have length 1. We assume that's the case in this
+  // implementation.
+
+  // Displacements are not implemented so return an error code if they are non-zero
+  if (displs[0] != 0) {
+    return MPI_ERR_INTERN;
+  }
+
+  MPIUNI_Memcpy(recvbuf, sendbuf, sendcounts[0]*sendtype, CHECK_FOR_MPI_IN_PLACE_DEST);
+  return MPI_SUCCESS;
+}
+
+int ESMC_MPI_Type_create_hvector(int count, int blocklength, MPI_Aint stride,
```

Comment:
Note that I ignore stride here: it seems like, for mpiuni, the important thing about type creation is to end up with a value that reflcts the size of the resulting type, and from some experimentation with this MPI function with openmpi, it seems like the MPI_Type_size of the result is independent of the stride.

However, this function is the one that I'm least certain of, so would be happy to have a second set of eyes on this.